### ISSUE TITLE
Unix: `exec` (`Switch-Process`) does not behave like an external (native) program syntactically, even though it should

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#24326]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Because of how `exec` is implemented - as a function - it doesn't behave like an external (native) program with respect to its invocation syntax (native globbing, tilde expansion, treating `,` literally, passing `--` through, ...)

This is unexpected and inconsistent with the `exec` builtin of POSIX shells it is meant to emulate.

Note: 

* **All the following repros _close the current session_  after waiting for <kbd>Enter</kbd>** (and must therefore each be submitted in a new session).

```powershell
# On a Unix-like platform

# !! Prints "~" verbatim.
exec sh -c 'echo "$@"; read' - ~

# !! Drops "--"
exec sh -c 'echo "$@"; read' - one -- two

# !! Fails with "Cannot convert 'System.Object[]' to the type 'System.String' required by parameter 'WithCommand'"
exec sh -c 'echo "$@"; read' - one,two
```

### Expected behavior

```console
All commands should behave as they would without `exec`
```

### Actual behavior

```console
The commands do not act like native-program calls, and act as described in the source-code comments.
```

### Environment data

```powershell
PowerShell 7.5.0-preview.3 on Unix-like platforms.
```

**Comment by rhubarb-geek-nz** [comment]
> This is unexpected and inconsistent with the `exec` builtin of POSIX shells it is meant to emulate.

PowerShell is PowerShell, why should it have to behave like a POSIX shell?

If exec is implemented as a PowerShell function and abides by the PowerShell rules then all should be well. PowerShell functions not behaving like PowerShell functions would be unexpected and inconsistent with PowerShell.

**Comment by rhubarb-geek-nz** [comment]
> /usr/bin/exec does the same:

```
$ ls -ld /usr/bin/exec
ls: cannot access '/usr/bin/exec': No such file or directory
```

You have a `/usr/bin/exec`? It should be a built in command to the shell, like cd. I agree it is doing what is wanted, to call exec, my comment was observing that Switch-Process is odd because the one thing it is not doing is switching the process, everything else changes except the process.

**Comment by mklement0** [comment]
* The aim of the `exec` function is to provide compatibility with the `exec` builtin of POSIX shells (and therefore restricted to Unix-like platforms): replace the current process with a new one (while keeping the PID, ...).

* As such, the sole purpose of `exec` is to mediate invocation of _external programs_.

* It is counterintuitive / easy to forget / an annoyance that the current `exec` implementation changes the invocation syntax of the external program passed to it: if `sh -c 'echo "$@"; read' - ~` works as intended, so should `exec sh -c 'echo "$@"; read' - ~`; to put it differently: **placing `exec` before a working command line should not break it**.

  * The choice of _implementation method_ - a PowerShell _function_ - gets in the way of the expected behavior here. Perhaps needless to say, that choice should not affect the solution.
  * There is no easy solution, given that _some_ form of _in-process_ command is necessary:
    * Conceivably, a new `[NativeParserBinding()]` attribute to decorate a function's (empty) `param()` block with could signal to PowerShell that a given function should act like an external (native) program with respect to how its arguments are parsed.
    * This could also be leveraged by third-party wrapper functions around external programs.

---

Two asides:

* Indeed, `exec` by its very nature only makes sense as a _built-in_ (in-process command).
* The name `Switch-Process` is defensible, as you're clearly substituting a new process for the current one, the fact that it retains the original identity notwithstanding.

**Comment by 237dmitry** [comment]
> You have a /usr/bin/exec?

Sorry. This is builtin

**Comment by rhubarb-geek-nz** [comment]
> * as you're clearly substituting a new process for the current one, the fact that it retains the original identity notwithstanding.

I presume you never read [UNIX Network Programming](https://www.amazon.co.uk/UNIX-Network-Programming-Richard-Stevens/dp/0139498761). The process stays the same, the loaded binary program and memory image changes. The file-descriptors, the signals, everything maintained by the kernel is maintained. It is exactly the same process, if you were waiting on it with waitpid() or similar nothing has changed.

**Comment by mklement0** [comment]
PowerShell has a limited number of approved verbs, and, among them, `Switch` is the best fit.
While conceiving of the current process as getting switched / replaced / swapped may not be fully technically accurate, pragmatically speaking it is good enough for the purpose at hand. 

Discussing this further here is a distraction; if you feel strongly that the name should change (which at this point would mean: introduce an _alias_), please open a separate issue (though I doubt you'll find a better fit than `Switch`).

As for what motivated the introduction of `exec`: I've added a quote from the docs to my previous comment.

**Comment by rhubarb-geek-nz** [comment]
Regarding the placement of exec. PowerShell is not a POSIX shell, its parsing and command structure is different. I don't understand the need to make PowerShell behave like bash in command parsing. If you want to use bash, then use bash. Otherwise we will have to turn exec into a keyword that only works on POSIX. The parsing rules and argument passing for functions are well founded. There is no other place in PowerShell where you can just add a word before a command and not expect it to change the whole interpretation.

**Comment by mklement0** [comment]
PowerShell is indeed not a POSIX shells, and fundamental differences will always remain, even in the context of `exec`.

However, PowerShell tries to be a relatively good citizen among POSIX shells with respect to supporting _running via shebang lines_ and _acting as a user's default shell_, and has made numerous changes to accommodate that (switching from `-Command` to `-File` as the default CLI parameter, supporting `-l` to act as a login shell, adding `-i`,  ... - see #3743 for what has and hasn't been implemented.

`exec` falls into that category (see #4691 for the backstory), and while the syntax cannot be expected to be fully POSIX-compatible, it is reasonable for it to behave _like_ an external program in terms of PowerShell's parsing, even though it technically cannot be. 

The problem may not often arise, but having to situationally rewrite working external-program command lines just to make them work with `exec` (to deal with differences in handling of `~`, globbing, `--`, `,`, ..., as mentioned), and just because of how `exec` happens to be implemented, strikes me as unacceptable, especially given that the need for such rewrites means that external utilities targeting the default shell are likelier to break the call due to behavior that only makes sense for PowerShell-native calls / lacks the compatibility accommodations.


As an aside: A conceptually related debate in the context of supporting command-scoped environment variables, possibly with POSIX syntax (`FOO=1 BAR=2 some-utility ...`): https://github.com/PowerShell/PowerShell/issues/3316#issuecomment-2356341169

**Comment by rhubarb-geek-nz** [comment]
PowerShell has much more in common with Perl and Python, both of which support shebang and both of which support exec through library calls as noted early.

The process model used by PowerShell is completely different to Bourne-based shells, for instance scripts are run in process rather than in a new process per script.

I don't share your opinion that exec falls into this category. The mechanism already exists (Switch-Process) but the idea of not being able to use bash syntax in a PowerShell is somehow unacceptable to be bizarre.

Is invoking exec supported in PowerShell, Perl, Python? Yes.
Is the syntax different to bash? Yes, just like absolutely everything else.

**Comment by mklement0** [comment]
Your observations about kinship with other languages and the process model strike me as irrelevant to this discussion.

> I don't share your opinion that exec falls into this category.

That's not my opinion; syntax compatibility with POSIX shells (to the extent possible, see below) is the rationale stated in the docs (and  #4691) for having introduced an `exec` command in PowerShell. 

> that not being able to use bash syntax in a PowerShell is somehow unacceptable to be bizarre.

No one has asked for that.
The argument is primarily one about _internal consistency_ in the PowerShell implementation of `exec`, from which the extent of the POSIX compatibility implicitly benefits; _full syntax compatibility with POSIX shell is neither possible nor desired_; e.g., `pwsh` supports a `-c` CLI parameter, like POSIX shells, but the argument must use PowerShell syntax.

The point is that making `exec` act _like an external program_ in terms of how PowerShell parses its arguments, _using regular PowerShell syntax for external programs_:
* minimizes the syntax differences between PowerShell and POSIX shells, so that an `exec` command submitted by an external utility that assumes a POSIX shell is less likely to break
* avoids extra, unexpected, obscure work and prevents bugs when it is called from PowerShell

Neither of these problems may arise often in practice, but I think they're worth solving, not least because something like a `[NativeParserBinding()]` attribute (or perhaps `[PSNativeCommandArgumentParsing()]`) could benefit other use cases too (e.g. #21208 and possibly the aforementioned #3316)

**Comment by rhubarb-geek-nz** [comment]
What is the scenario you are trying to cover here? 

> minimizes the syntax differences between PowerShell and POSIX shells, so that an exec command submitted by an external utility that assumes a POSIX shell is less likely to break

If you have installed PowerShell as your default shell, and your SHELL environment is `/usr/bin/pwsh` then you will get what you asked for. No existing code expects a non-POSIX shell to be the default shell. I am not sure why `exec` is a special case.

**Comment by rhubarb-geek-nz** [comment]
I will ask the question another way round.

Should everything in Windows work if I set COMSPEC to c:\windows\system32\bash.exe. c:\Windows\system32\WindowsPowerShell\v1.0\powershell.exe or c:\Program Files\PowerShell\7\pwsh.exe?

**Comment by SteveL-MSFT** [comment]
@mklement0 there's a technical limitation to make `Switch-Process` work exactly the same as the `exec` built-in for bash.  As you can see, the [cmdlet implementation](https://github.com/PowerShell/PowerShell/blob/c191efe890c74e6227a803905d7eb342728484fc/src/System.Management.Automation/engine/Modules/SwitchProcessCommand.cs#L22) relies on `ValueFromRemainingArguments` so all the normal PowerShell parameter binding rules apply and prevent what you want.

The original goal of `Switch-Process` was to accommodate some known problems applications had running when PowerShell was default shell as they expected `exec` to exist.  Do you have a real world common example that we need to look at more closely?

**Comment by mklement0** [comment]
@SteveL-MSFT:

> there's a technical limitation to make Switch-Process work exactly the same as the exec built-in for bash

As for the limitation with respect to _stream redirections_: understood, but this in't what this issue is about.

> As you can see, the [cmdlet implementation](https://github.com/PowerShell/PowerShell/blob/c191efe890c74e6227a803905d7eb342728484fc/src/System.Management.Automation/engine/Modules/SwitchProcessCommand.cs#L22) relies on ValueFromRemainingArguments so all the normal PowerShell parameter binding rules apply and prevent what you want.

Indeed, and solving this problem with a `[PSNativeCommandArgumentParsing()]` attribute that would allow (at least) _non-advanced_ PowerShell _functions_  to act _as if they were external programs_ with respect to parameter binding is what this issue comes down to (for _advanced_ functions as well as _cmdlets_,  the attribute would have to additionally disable all _common parameters_ in order to work as expected).

As noted, such an attribute would enable creating user-defined wrappers for external programs too.

> Do you have a real world common example that we need to look at more closely?

The examples in the initial post demonstrate methodically where the PowerShell-native argument parsing diverges from external-program argument parsing.

While I have no specific real-world examples, it's also not hard to imagine how uses of `exec` - both by outside callers via `pwsh -c` and PowerShell-internally - can run afoul of this divergence.
